In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/imakash3011/customer-personality-analysis/marketing_campaign.csv


# Customer Personality Analysis - Data Cleaning

## Objective
The goal of this project is to clean the Customer Personality Analysis dataset and transform it into a reliable, analysis-ready dataset.

The cleaning process includes:
- Data quality assessment
- Missing value treatment
- Duplicate removal
- Data standardization
- Outlier detection
- Data type correction
- Before vs After comparison
- Exporting the cleaned dataset

In [2]:
import pandas as pd

In [3]:
df = pd.read_csv("/kaggle/input/datasets/imakash3011/customer-personality-analysis/marketing_campaign.csv", sep='\t')
df

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1957,Graduation,Single,58138.0,0,0,04-09-2012,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1954,Graduation,Single,46344.0,1,1,08-03-2014,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1965,Graduation,Together,71613.0,0,0,21-08-2013,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1984,Graduation,Together,26646.0,1,0,10-02-2014,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1981,PhD,Married,58293.0,1,0,19-01-2014,94,173,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1967,Graduation,Married,61223.0,0,1,13-06-2013,46,709,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1946,PhD,Together,64014.0,2,1,10-06-2014,56,406,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1981,Graduation,Divorced,56981.0,0,0,25-01-2014,91,908,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1956,Master,Together,69245.0,0,1,24-01-2014,8,428,...,3,0,0,0,0,0,0,3,11,0


In [4]:
print(df.info)

<bound method DataFrame.info of          ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
0      5524        1957  Graduation         Single  58138.0        0   
1      2174        1954  Graduation         Single  46344.0        1   
2      4141        1965  Graduation       Together  71613.0        0   
3      6182        1984  Graduation       Together  26646.0        1   
4      5324        1981         PhD        Married  58293.0        1   
...     ...         ...         ...            ...      ...      ...   
2235  10870        1967  Graduation        Married  61223.0        0   
2236   4001        1946         PhD       Together  64014.0        2   
2237   7270        1981  Graduation       Divorced  56981.0        0   
2238   8235        1956      Master       Together  69245.0        0   
2239   9405        1954         PhD        Married  52869.0        1   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumWebVisitsMonth  \
0            0  04-09-2012    

In [5]:
#Checking for null values
print(df.shape)
print(df.isnull().sum())

(2240, 29)
ID                      0
Year_Birth              0
Education               0
Marital_Status          0
Income                 24
Kidhome                 0
Teenhome                0
Dt_Customer             0
Recency                 0
MntWines                0
MntFruits               0
MntMeatProducts         0
MntFishProducts         0
MntSweetProducts        0
MntGoldProds            0
NumDealsPurchases       0
NumWebPurchases         0
NumCatalogPurchases     0
NumStorePurchases       0
NumWebVisitsMonth       0
AcceptedCmp3            0
AcceptedCmp4            0
AcceptedCmp5            0
AcceptedCmp1            0
AcceptedCmp2            0
Complain                0
Z_CostContact           0
Z_Revenue               0
Response                0
dtype: int64


The income column has 24 null values

In [6]:
#Looking for duplicates
print(df.duplicated().sum())

#Checking the datatypes
print(df.dtypes)

0
ID                       int64
Year_Birth               int64
Education               object
Marital_Status          object
Income                 float64
Kidhome                  int64
Teenhome                 int64
Dt_Customer             object
Recency                  int64
MntWines                 int64
MntFruits                int64
MntMeatProducts          int64
MntFishProducts          int64
MntSweetProducts         int64
MntGoldProds             int64
NumDealsPurchases        int64
NumWebPurchases          int64
NumCatalogPurchases      int64
NumStorePurchases        int64
NumWebVisitsMonth        int64
AcceptedCmp3             int64
AcceptedCmp4             int64
AcceptedCmp5             int64
AcceptedCmp1             int64
AcceptedCmp2             int64
Complain                 int64
Z_CostContact            int64
Z_Revenue                int64
Response                 int64
dtype: object


No duplicates
ID, Year_Birth and Dt_customer columns are in the wrong format

In [7]:
before_summary = {
    "Rows": df.shape[0],
    "Columns": df.shape[1],
    "Missing Values": df.isnull().sum().sum(),
    "Duplicate Rows": df.duplicated().sum(),
    "Incorrect Data Types": 3
}
print(before_summary)

{'Rows': 2240, 'Columns': 29, 'Missing Values': np.int64(24), 'Duplicate Rows': np.int64(0), 'Incorrect Data Types': 3}


Before cleaning the dataset, there existed 2240 rows, 29 columns, 24 missing values no duplicate rows, 3 columns with incorrect data type

In [8]:
#Converting Columns to correct format
df['ID'] = df['ID'].astype(str)
df['Year_Birth'] = pd.to_datetime(df['Year_Birth'],dayfirst=True).dt.year
df['Dt_Customer'] = pd.to_datetime(df['Dt_Customer'],dayfirst=True)
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2240 entries, 0 to 2239
Data columns (total 29 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID                   2240 non-null   object        
 1   Year_Birth           2240 non-null   int32         
 2   Education            2240 non-null   object        
 3   Marital_Status       2240 non-null   object        
 4   Income               2216 non-null   float64       
 5   Kidhome              2240 non-null   int64         
 6   Teenhome             2240 non-null   int64         
 7   Dt_Customer          2240 non-null   datetime64[ns]
 8   Recency              2240 non-null   int64         
 9   MntWines             2240 non-null   int64         
 10  MntFruits            2240 non-null   int64         
 11  MntMeatProducts      2240 non-null   int64         
 12  MntFishProducts      2240 non-null   int64         
 13  MntSweetProducts     2240 non-nul

In [9]:
#Filling the null records
median_income = df['Income'].median()
df['Income'] = df['Income'].fillna(median_income)

Income contained missing values. Since income distributions are often right-skewed and sensitive to extreme values, the median was used for imputation instead of the mean.

In [10]:
#Checking for unique values
print(df['Marital_Status'].unique())

['Single' 'Together' 'Married' 'Divorced' 'Widow' 'Alone' 'Absurd' 'YOLO']


We see that the Marital status column consists of various unique values most of which are unnecessary and can be accomodated in already existing names which is what we'll try to achieve in the next cell

In [11]:
df['Marital_Status'] = df['Marital_Status'].replace({
    'Absurd':'Other',
    'YOLO': 'Other',
    'Together': 'Married',
    'Alone': 'Single',
    'Widow': 'Single'
})
print(df['Marital_Status'].unique())

['Single' 'Married' 'Divorced' 'Other']


The values of the Marital Status column have been trimmed down to: Single, Married, Divorced, Other

In [12]:
#Outlier Detection
numeric_cols = df.select_dtypes(include=np.number).columns

#Loop through them using the IQR method.
for col in numeric_cols:

    Q1 = df[col].quantile(.25)
    Q3 = df[col].quantile(.75)

    IQR = Q3 - Q1

    lower = Q1 - 1.5*IQR
    upper = Q3 + 1.5*IQR

    outliers = ((df[col] < lower) | (df[col] > upper)).sum()

    print(col, outliers)

Year_Birth 0
Income 8
Kidhome 0
Teenhome 0
Recency 0
MntWines 35
MntFruits 227
MntMeatProducts 175
MntFishProducts 223
MntSweetProducts 248
MntGoldProds 207
NumDealsPurchases 86
NumWebPurchases 4
NumCatalogPurchases 23
NumStorePurchases 0
NumWebVisitsMonth 8
AcceptedCmp3 163
AcceptedCmp4 167
AcceptedCmp5 163
AcceptedCmp1 144
AcceptedCmp2 30
Complain 21
Z_CostContact 0
Z_Revenue 0
Response 334


### Outlier Detection

Outliers were identified using the Interquartile Range (IQR) method. A few columns, especially those related to customer income and product spending, contained values outside the expected range. After reviewing these values, I decided to keep them because they appear to reflect genuine customer behaviour rather than data entry mistakes. For example, customers with very high incomes or spending habits are realistic and could provide valuable insights during analysis.

Binary columns such as `AcceptedCmp1`, `Response`, and `Complain` were not treated as outliers because they only contain two possible values (0 and 1), making the IQR method unsuitable for those variables.


In [13]:
after_summary = {
    "Rows": df.shape[0],
    "Columns": df.shape[1],
    "Missing Values": df.isnull().sum().sum(),
    "Duplicate Rows": df.duplicated().sum(),
    "Incorrect Data Types": 0
}
print(after_summary)

{'Rows': 2240, 'Columns': 29, 'Missing Values': np.int64(0), 'Duplicate Rows': np.int64(0), 'Incorrect Data Types': 0}


After cleaning the dataset, there are 2240 rows, 29 columns,no missing values, no duplicate rows, 0 columns with incorrect data type.
Comparison below

In [14]:
summary = pd.DataFrame({
    "Before Cleaning": before_summary,
    "After Cleaning": after_summary
})

print(summary)

                      Before Cleaning  After Cleaning
Rows                             2240            2240
Columns                            29              29
Missing Values                     24               0
Duplicate Rows                      0               0
Incorrect Data Types                3               0


In [15]:
#Save the cleaned dataset
df.to_csv('customer_personality_cleaned.csv', index=False)
df

,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Z_CostContact,Z_Revenue,Response
0,5524,1970,Graduation,Single,58138.0,0,0,2012-09-04,58,635,...,7,0,0,0,0,0,0,3,11,1
1,2174,1970,Graduation,Single,46344.0,1,1,2014-03-08,38,11,...,5,0,0,0,0,0,0,3,11,0
2,4141,1970,Graduation,Married,71613.0,0,0,2013-08-21,26,426,...,4,0,0,0,0,0,0,3,11,0
3,6182,1970,Graduation,Married,26646.0,1,0,2014-02-10,26,11,...,6,0,0,0,0,0,0,3,11,0
4,5324,1970,PhD,Married,58293.0,1,0,2014-01-19,94,173,...,5,0,0,0,0,0,0,3,11,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2235,10870,1970,Graduation,Married,61223.0,0,1,2013-06-13,46,709,...,5,0,0,0,0,0,0,3,11,0
2236,4001,1970,PhD,Married,64014.0,2,1,2014-06-10,56,406,...,7,0,0,0,1,0,0,3,11,0
2237,7270,1970,Graduation,Divorced,56981.0,0,0,2014-01-25,91,908,...,6,0,1,0,0,0,0,3,11,0
2238,8235,1970,Master,Married,69245.0,0,1,2014-01-24,8,428,...,3,0,0,0,0,0,0,3,11,0
